### PR TITLE
[dattri] Add huggingface example (GPT-2 + WikiText-2) and related bug fix

### PR DESCRIPTION
**Pull request by TheaperDeng**
### 1. Motivation and Context

- Models from huggingface `transformers` is widely used and some additional fix should be included in `dattri`.
- A larger experiment is added (GPT-2 + WikiText-2) to `experiments/`

### 2. Summary of the change

Experiment added:
- An example GPT-2 + WikiText-2 is added to `experiments/gpt2_wikitext`. Please check https://github.com/TheaperDeng/dattri-public/tree/gpt2_wiki2_hf/experiments/gpt2_wikitext for more detail.

Example added
- An example using Dropout Ensemble in https://github.com/TheaperDeng/dattri-public/blob/gpt2_wiki2_hf/examples/pretrained_benchmark/trak_dropout_lds.py

API change (backlog supported, no need for user code change):
- Add `checkpoints_load_func` to `AttributionTask`, The parameter is used for some model that has special loading strategies, e.g., huggingface model. A typical example for huggingface model is

```python
def checkpoints_load_func(model, checkpoint):
    model = AutoModelForCausalLM.from_pretrained(checkpoint).cuda()
    model.eval()
    return model

task = AttributionTask(loss_func=f, model=model,
                       checkpoints=checkpoints,
                       checkpoints_load_func=checkpoints_load_func)  # if not specific, a default one will be applied.
```

Bug fix:
- For `TRAKAttributor`
   - The scaling number calculation is changed to a much more efficient implementation.
   - The "pre-gradient" calculation is changed to support non-tuple/list data format (e.g., dict, that is widely used for models in `transformers`
   - Change batch size calculation method to the shape[0] of gradient rather than input data

TODO:
- Better "pre-gradient" calculation implementation for TRAK and other attributors. (has been recorded to source code).

### 3. What tests have been added/updated for the change?
- [ ] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.

### CODE REVIEW

File: dattri/algorithm/trak.py [hunk 1]
```diff
@@ -84,6 +89,7 @@ def m(params, image_label_pair):
         self.correct_probability_func = vmap(
             correct_probability_func,
             in_dims=(None, 0),
+            randomness="different",
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:92]: I am wondering why we suddenly need this? I see there are also other places adding this. Does this affect the performance much?

File: dattri/algorithm/trak.py [hunk 2]
```diff
@@ -127,14 +133,22 @@ def cache(
                 desc="calculating gradient of training set...",
                 leave=False,
             ):
-                train_batch_data = tuple(data.to(self.device) for data in train_data)
+                # TODO: reorganize the data pre-grad processing.
+                if isinstance(train_data, (tuple, list)):
+                    train_batch_data = tuple(
+                        data.to(self.device) for data in train_data
+                    )
+                else:
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:141]: per your description, so if `train_data` is not a tuple of a list, then it must be a dict?
And why if `train_data` is a dict, then we don't need to put data onto `self.device`?
In addition, do we need to restrict the type of `train_data` here?

File: dattri/algorithm/trak.py [hunk 3]
```diff
@@ -239,20 +253,26 @@ def attribute(  # noqa: PLR0912
...
-                        data.to(self.device) for data in train_data
-                    )
+                    # TODO: reorganize the data pre-grad processing.
+                    if isinstance(train_data, (tuple, list)):
+                        train_batch_data = tuple(
+                            data.to(self.device) for data in train_data
+                        )
+                    else:
+                        train_batch_data = train_data
+
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:263]: the same issue as above

File: dattri/algorithm/trak.py [hunk 4]
```diff
@@ -282,15 +302,19 @@ def attribute(  # noqa: PLR0912
                 desc="calculating gradient of test set...",
                 leave=False,
             ):
-                test_batch_data = tuple(data.to(self.device) for data in test_data)
+                # TODO: reorganize the data pre-grad processing.
+                if isinstance(test_data, (tuple, list)):
+                    test_batch_data = tuple(data.to(self.device) for data in test_data)
+                else:
+                    test_batch_data = test_data
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:309]: the same issue as above

File: dattri/task.py [hunk 5]
```diff
@@ -75,8 +95,19 @@ def f(params, data):
                     loss = nn.CrossEntropyLoss()
                     yhat = torch.func.functional_call(model, params, image)
                     return loss(yhat, label)
+                ```
+            checkpoints_load_func (Callable): The checkpoint load function.
+                The input is optional, if not provided, the checkpoint load
+                function will be a default one using model.load_state_dict.
+                The parameter is used for some model that has special
```
Comment (tingwl0122) [inline dattri/task.py:102]: some models that have special...

File: dattri/task.py [hunk 6]
```diff
@@ -90,6 +121,11 @@ def f(params, data):
         self.original_target_func = target_func
         self.target_func = flatten_func(self.model)(target_func)
 
+        if checkpoints_load_func is None:
+            self.checkpoints_load_func = _default_checkpoint_load_func
+        else:
+            self.checkpoints_load_func = checkpoints_load_func
+
```
Comment (tingwl0122) [inline dattri/task.py:128]: do we need to be careful here? now it seems like we are either using default or huggingface loading.
Will there be any other type of loading?

File: dattri/task.py [hunk 7]
```diff
@@ -123,20 +167,14 @@ def _load_checkpoints(self, ckpt_idx: int) -> None:
...
-                self.model.load_state_dict(self.checkpoints[ckpt_idx])
+            self.model = self.checkpoints_load_func(
+                self.model,
+                self.checkpoints[ckpt_idx],
+            )
             self.current_checkpoint_idx = ckpt_idx
             self.named_parameters = {
                 k: p for k, p in self.model.named_parameters() if p.requires_grad
             }
-        self.model.eval()
```
Comment (tingwl0122) [inline dattri/task.py:139]: what is the original purpose of this line?

File: examples/pretrained_benchmark/trak_dropout_lds.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        score = attributor.attribute(
+            DataLoader(
+                model_details["test_dataset"],
+                batch_size=5000,
+                sampler=model_details["test_sampler"],
+            ),
+        )
+
+    lds_score = lds(score, groundtruth)[0]
+    print("lds:", torch.mean(lds_score[~torch.isnan(lds_score)]))
```
Comment (tingwl0122) [inline examples/pretrained_benchmark/trak_dropout_lds.py:87]: this example looks to me if we have validated its performance gain matches our expectation.

**Review by tingwl0122 — COMMENTED** [review]
I read through the files other than the `experiments/gpt2_wikitext` folder.
My only concern is that whether we need to unify this experiment into existing ones.
Does the original nanogpt resembles gpt2 used here?
If yes, then I guess it is just nanogpt + (ShakeSpeare or Wikitext)?

File: .github/workflows/examples_test.yml [hunk 9]
```diff
@@ -27,6 +27,8 @@ jobs:
         python examples/noisy_label_detection/trak_noisy_label.py --device cpu
         python examples/pretrained_benchmark/influence_function_lds.py --device cpu
         python examples/pretrained_benchmark/trak_loo.py --device cpu
+        sed -i 's/* 10/* 2/g' examples/pretrained_benchmark/trak_dropout_lds.py
```
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply .github/workflows/examples_test.yml:30]: This line is using `sed` to change the file (only during the test) and reduce the computational cost (ensemble number changed from 10 to 2). It will not affect the raw file.

File: dattri/algorithm/trak.py (hunk 1 above)
Comment (TheaperDeng) [reply dattri/algorithm/trak.py:92]: No, it does not affect the performance. But if we use dropout enabled during the computation, we need to state this `randomness="different“` or there will be error (defaultly, `randomness="error"`). Another choice is `randomness="same"`, this will affect the performance of dropout ensemble.

File: dattri/algorithm/trak.py (hunk 2 above)
Comment (TheaperDeng) [reply dattri/algorithm/trak.py:141]: I leave this to future PR. As I stated in TODO (`TODO: reorganize the data pre-grad processing.`). I will raise a PR for future developers to do this. I understand now the preprocessing is very weak but it seems too much for this PR.

File: dattri/algorithm/trak.py (hunk 3 above)
Comment (TheaperDeng) [reply dattri/algorithm/trak.py:263]: same

File: dattri/algorithm/trak.py (hunk 4 above)
Comment (TheaperDeng) [reply dattri/algorithm/trak.py:309]: same

File: dattri/task.py (hunk 6 above)
Comment (TheaperDeng) [reply dattri/task.py:128]: The logic of this code block is. If `checkpoints_load_func` is not explicitly provided, we will use `_default_checkpoint_load_func`. `checkpoints_load_func` is provided by users, it can be huggingface loading or any other things.

File: dattri/task.py (hunk 7 above)
Comment (TheaperDeng) [reply dattri/task.py:139]: When model is loaded, it will be changed to eval mode. Now it's in `_default_checkpoint_load_func`.

File: examples/pretrained_benchmark/trak_dropout_lds.py (hunk 8 above)
Comment (TheaperDeng) [reply examples/pretrained_benchmark/trak_dropout_lds.py:87]: Yes, it has been validated.

File: .github/workflows/examples_test.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (tingwl0122) [reply .github/workflows/examples_test.yml:30]: got it

File: dattri/algorithm/trak.py (hunk 1 above)
Comment (tingwl0122) [reply dattri/algorithm/trak.py:92]: Got it. Just curious, so if we state `randomness="same"`, there will only be one dropout version basically?

File: dattri/algorithm/trak.py (hunk 2 above)
Comment (tingwl0122) [reply dattri/algorithm/trak.py:141]: Make sense

File: dattri/task.py (hunk 6 above)
Comment (tingwl0122) [reply dattri/task.py:128]: Oh got it. sorry for the misunderstanding

File: dattri/task.py (hunk 7 above)
Comment (tingwl0122) [reply dattri/task.py:139]: got it

**Review by tingwl0122 — APPROVED** [review]
LGTM now
so should we raise an issue about the gpt2+wiki setting reorganization?

File: dattri/algorithm/trak.py (hunk 1 above)
Comment (TheaperDeng) [reply dattri/algorithm/trak.py:92]: I think it will be one dropout version for each independent checkpoint. So there is no meaning to use dropout ensemble anymore.

**Comment by TheaperDeng** [comment]
> LGTM now so should we raise an issue about the gpt2+wiki setting reorganization?

Yes, I will raise it after merging the PR.
